tests/phpunit/includes/test-wc-gateway-amazon-payments-advanced.php

```diff
@@ -0,0 +1,147 @@
+ */
+class WC_Gateway_Amazon_Payments_Advanced_Test extends WP_UnitTestCase {
+
+	/**
+	 * Store an instance of our Gateway being tested.
+	 *
+	 * @var ?WC_Gateway_Amazon_Payments_Advanced
+	 */
+	protected static $gateway;
+
+	/**
+	 * Set up our DB before tests.
+	 *
+	 * @return void
+	 */
+	public static function set_up_before_class() : void {
+		parent::set_up_before_class();
+		self::$gateway = new WC_Gateway_Amazon_Payments_Advanced();
+		update_option( 'woocommerce_currency', 'EUR' );
+		update_option( 'woocommerce_amazon_payments_new_install', WC_AMAZON_PAY_VERSION );
+		update_option( 'amazon_api_version', 'V2' );
+	}
+
+	/**
+	 * Restore the DB's state after tests.
+	 *
+	 * @return void
+	 */
+	public static function tear_down_after_class() : void {
+		parent::tear_down_after_class();
+		self::$gateway = null;
+		delete_option( 'woocommerce_currency' );
+		delete_option( 'woocommerce_amazon_payments_advanced_settings' );
+		delete_option( 'woocommerce_amazon_payments_new_install' );
+		delete_option( 'amazon_api_version' );
+		delete_option( WC_Amazon_Payments_Advanced_Merchant_Onboarding_Handler::KEYS_OPTION_PRIVATE_KEY );
+	}
+
+	/**
+	 * Test that gateway is not available.
+	 *
+	 * @return void
+	 */
+	public function test_is_not_available() : void {
+		$this->assertFalse( self::$gateway->is_available() );
+	}
+
+	/**
+	 * Test that gateway is available.
+	 *
+	 * @return void
+	 */
+	public function test_is_available() : void {
+		self::update_plugin_settings();
+		$this->assertTrue( self::$gateway->is_available() );
+	}
+
+	/**
+	 * Test happy path of the express gateway.
+	 *
+	 * @return void
+	 */
+	public function test_express_process_payment() : void {
+		$checkout_session_key = apply_filters( 'woocommerce_amazon_pa_checkout_session_key', 'amazon_checkout_session_id' );
+		WC()->session->set( $checkout_session_key, 'TEST_CHECKOUT_SESSION_ID' );
+		WC()->session->save_data();
+
+		$order = WC_Helper_Order::create_order( self::$gateway->id );
+
+		$mock_gateway = new WC_Mocker_Gateway_Amazon_Payments_Advanced();
+
+		$this->assertEquals(
+			array(
+				'result'   => 'success',
+				'redirect' => 'https://amazon.unit.tests/',
+			),
+			$mock_gateway->process_payment( $order->get_id() )
+		);
+	}
+
+	/**
+	 * Test happy path of classic gateway.
+	 *
+	 * @return void
+	 */
+	public function test_classic_process_payment() : void {
+		$checkout_session_key = apply_filters( 'woocommerce_amazon_pa_checkout_session_key', 'amazon_checkout_session_id' );
+		WC()->session->set( $checkout_session_key, null );
+		WC()->session->save_data();
+
+		$order_total = 100;
+
+		$order = WC_Helper_Order::create_order( self::$gateway->id, 1, $order_total );
+
+		$mock_gateway = new WC_Mocker_Gateway_Amazon_Payments_Advanced( $order_total );
+
+		remove_action( 'woocommerce_after_checkout_validation', array( wc_apa()->get_gateway(), 'classic_validation' ), 10 );
```

Comment:
From the test case itself, it's not clear why to drop it. Can we add a comment perhaps?

Reply:
It seems like its not needed after re testing this. Probably an error initially made me think that was the problem. Stopped removing the action.